## Group SQL Practice

_Author: Matt Brems & Katie Sylvia_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [63]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

In [64]:
engine

Engine(postgresql://analytics_student:***@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database)

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [65]:
sql = """
SELECT COUNT(store)
FROM stores

"""

pd.read_sql_query(sql, engine)

count
0   1973

In [66]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(store)
FROM stores

"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: There are 1,973 rows in the stores table.

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [67]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT name
FROM stores
ORDER BY name
LIMIT 4
"""

pd.read_sql_query(sql, engine)

name
0  218 Fuel Express & Chubby's Liquor
1          3047 Cub Foods / Iowa City
2               3060 Cub Foods / Ames
3         3061 Cub Foods / Sioux City

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [68]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, bottle_qty, vendor
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor
LIMIT 10

"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty                     vendor
0     $8.99          12  Adamba Imports Int'l Inc.
1    $28.50          30      A Hardy / U.S.A. Ltd.
2    $28.50           1      A Hardy / U.S.A. Ltd.
3    $28.50           3      A Hardy / U.S.A. Ltd.
4    $28.50           6      A Hardy / U.S.A. Ltd.
5    $28.50           6      A Hardy / U.S.A. Ltd.
6    $28.50           6      A Hardy / U.S.A. Ltd.
7    $28.50           6      A Hardy / U.S.A. Ltd.
8    $28.50           3      A Hardy / U.S.A. Ltd.
9    $28.50           6      A Hardy / U.S.A. Ltd.

**Prompt 3 Answer**: 
- 8.99, 12, Adamba Imports Int'l 
- 28.50, 6, A Hardy / USA Ltd. 
- 28.50, 6, A Hardy / USA Ltd.


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [69]:

sql = """
SELECT *
FROM sales
WHERE date BETWEEN '2015-02-01' AND '2015-02-28';
"""
pd.read_sql_query(sql, engine)

date convenience_store  store county_number         county  \
0      2015-02-25                 Y   4932             7     Black Hawk   
1      2015-02-26              None   5012             8          Boone   
2      2015-02-26              None   5012             8          Boone   
3      2015-02-26              None   5012             8          Boone   
4      2015-02-26              None   5012             8          Boone   
...           ...               ...    ...           ...            ...   
158276 2015-02-17              None   3963            78  Pottawattamie   
158277 2015-02-17              None   3963            78  Pottawattamie   
158278 2015-02-17              None   3963            78  Pottawattamie   
158279 2015-02-17              None   3963            78  Pottawattamie   
158280 2015-02-17              None   3963            78  Pottawattamie   

       category                     category_name vendor_no  \
0       1062200  PUERTO RICO & VIRGIN ISLANDS RUM       434   
1       1031080                    80 PROOF VODKA       260   
2       1031200                    FLAVORED VODKA       380   
3       1031200                    FLAVORED VODKA       380   
4       1062200  PUERTO RICO & VIRGIN ISLANDS RUM        35   
...         ...                               ...       ...   
158276  1031080                    80 PROOF VODKA       205   
158277  1031200                    FLAVORED VODKA       205   
158278  1031200                    FLAVORED VODKA       205   
158279  1031200                    FLAVORED VODKA       205   
158280  1031200                    FLAVORED VODKA       205   

                           vendor   item                   description  pack  \
0                  Luxco-St Louis  45277           Paramount White Rum    12   
1                 Diageo Americas  38008     Smirnoff Vodka 80 Prf Pet     6   
2       Phillips Beverage Company  41783  Uv Blue Raspberry Vodka Mini     6   
3       Phillips Beverage Company  41857      Uv Cherry Red Vodka Mini     6   
4            Bacardi U.S.A., Inc.  43125      Bacardi Superior Rum Pet    12   
...                           ...    ...                           ...   ...   
158276       E AND J GALLO WINERY  36969           New Amsterdam Vodka    12   
158277       E AND J GALLO WINERY  40185          New Amsterdam Orange    24   
158278       E AND J GALLO WINERY  40186          New Amsterdam Orange    12   
158279       E AND J GALLO WINERY  40191       New Amsterdam Pineapple    24   
158280       E AND J GALLO WINERY  40192       New Amsterdam Pineapple    24   

        liter_size state_btl_cost btl_price  bottle_qty   total  
0             1000          $4.34     $6.51          12   78.12  
1             1750         $14.75    $22.13           6  132.78  
2              500          $4.89     $7.34           6   44.04  
3              500          $4.89     $7.34           6   44.04  
4              750          $8.00    $12.00           2   24.00  
...            ...            ...       ...         ...     ...  
158276         750          $6.90    $10.35           1   10.35  
158277         375          $3.50     $5.25           3   15.75  
158278         750          $6.90    $10.35           2   20.70  
158279         200          $2.24     $3.36           2    6.72  
158280         375          $3.50     $5.25           2   10.50  

[158281 rows x 17 columns]

**Prompt 4 Answer**: $21,295,350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [70]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(vendor) AS ttl_transaction, SUM(bottle_qty) AS bottles_sold, SUM(total) AS ttl_rev, vendor
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY ttl_transaction
DESC
LIMIT 1;
"""

pd.read_sql_query(sql, engine)

ttl_transaction  bottles_sold     ttl_rev                            vendor
0            42338        403450  7880365.88  Pernod Ricard USA/Austin Nichols

**Prompt 5 Answer**: Pernod Ricard has the most transactios at 42,338. With 403,450 bottles sold.

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [71]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price - state_btl_cost AS markup, vendor, description, convenience_store
FROM sales
ORDER by markup
DESC;



"""

pd.read_sql_query(sql, engine)

markup                            vendor  \
0        $2,900.00          Cedar Ridge VineyardsLLC   
1        $2,900.00          Cedar Ridge VineyardsLLC   
2          $799.60  Pernod Ricard USA/Austin Nichols   
3          $799.60  Pernod Ricard USA/Austin Nichols   
4          $799.60  Pernod Ricard USA/Austin Nichols   
...            ...                               ...   
3049908      $0.49  Pernod Ricard USA/Austin Nichols   
3049909      $0.49  Pernod Ricard USA/Austin Nichols   
3049910      $0.49  Pernod Ricard USA/Austin Nichols   
3049911      $0.28               Bacardi U.S.A. Inc.   
3049912      $0.24         Monsieur Henri Wines Ltd.   

                                  description convenience_store  
0            Cedar Ridge Barrel Proof Bourbon              None  
1            Cedar Ridge Barrel Proof Bourbon              None  
2        Absolut w/ Zing Zang Bloody Mary Mix              None  
3        Absolut w/ Zing Zang Bloody Mary Mix              None  
4        Absolut w/ Zing Zang Bloody Mary Mix              None  
...                                       ...               ...  
3049908               Seagram's Extra Dry Gin              None  
3049909               Seagram's Extra Dry Gin              None  
3049910               Seagram's Extra Dry Gin                 Y  
3049911          Bacardi Black Razz Miniature              None  
3049912                     Gran Gala Liqueur              None  

[3049913 rows x 4 columns]

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [144]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT store, ROUND(AVG((btl_price - state_btl_cost) / state_btl_cost)* 100) AS avg_markup_perc
FROM sales
GROUP BY store
ORDER BY avg_markup_perc 
DESC;
"""

pd.read_sql_query(sql, engine)

store  avg_markup_perc
0      4013             53.0
1      3833             52.0
2      4266             52.0
3      4024             52.0
4      3713             51.0
...     ...              ...
1347   2595             50.0
1348   2596             50.0
1349   2597             50.0
1350   2599             50.0
1351   2600             50.0

[1352 rows x 2 columns]

**Prompt 7 Answer**: Store 4013 has the largest average markup at 53%

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [151]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sales.store,
    ROUND(AVG((btl_price - state_btl_cost)/state_btl_cost)*100) as avg_markup_perc,
    ROUND(MIN((btl_price - state_btl_cost)/state_btl_cost)*100) as min_markup_perc,
    ROUND(MAX((btl_price - state_btl_cost)/state_btl_cost)*100) as max_markup_perc,
    stores.name, 
    stores.store_address
FROM sales, stores
WHERE sales.store = stores.store 
GROUP BY sales.store, stores.name, stores.store_address
ORDER BY avg_markup_perc
DESC
LIMIT 5;

"""

pd.read_sql_query(sql, engine)

store  avg_markup_perc  min_markup_perc  max_markup_perc  \
0   4013             53.0             50.0            671.0   
1   4266             52.0             50.0            671.0   
2   3833             52.0             50.0            671.0   
3   4024             52.0             50.0            671.0   
4   2668             51.0             50.0             69.0   

                         name  \
0      Wal-Mart 0841 / Tipton   
1  Wal-Mart 1683 / Shenandoah   
2    Wal-Mart 3394 / Atlantic   
3  Wal-Mart 1546 / Iowa Falls   
4    Hy-Vee Food Store / Leon   

                                       store_address  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  
1  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  
2  1905 East 7th St\nAtlantic, IA 500220000\n(41....  
3  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  
4  1004 W 1st St\nLeon, IA 50144\n(40.73967779500...

**Prompt 8 Answer**: 
- Walmart
- Walmart
- Walmart
- Walmart
- Hy-Vee Food Store/Leon


### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [122]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT
    counties.county,
    (SUM(sales.bottle_qty) / counties.population) AS btl_per_cap,
    ROUND(AVG(sales.liter_size), 1) AS avg_size
FROM counties
JOIN sales ON sales.county = counties.county
GROUP BY counties.county, counties.population
ORDER BY btl_per_cap DESC
LIMIT 9;
"""

pd.read_sql_query(sql, engine)

county  btl_per_cap  avg_size
0      Dickinson           22     974.2
1           Polk           15     891.9
2     Black Hawk           15     861.0
3    Cerro Gordo           14     986.4
4          Scott           13     869.9
5        Johnson           13     895.2
6           Linn           12     885.1
7  Pottawattamie           11     890.9
8        Kossuth           11    1040.5

**Prompt 9 Answer**: Kossuth sells the largest bottle size. 

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [133]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*) AS mismatched_transactions
FROM sales
WHERE total::numeric <> (btl_price * bottle_qty)::numeric;
"""

pd.read_sql_query(sql, engine)

mismatched_transactions
0                        0

**Prompt 10 Answer**: 0 transactions